### PR TITLE
Temperature API

### PR DESCRIPTION
**Pull request by alexandruCarp**
### Pull Request Overview

This pull request adds:
- a Temperature API, similar to the [libtock-c version](https://github.com/tock/libtock-c/blob/master/libtock/temperature.h). The API provides a function that simulates a synchronous temperature reading and also functions to initiate a reading, register/ unregister a listener and check for driver existence. 
- an example app which samples the temperature sensor every 2 seconds and prints the measured temperature.
- a `fake` temperature driver, for testing purposes.
- unit tests for the fake driver and for the user space library

### Testing Strategy

This pull request was tested with the unit tests and by running the example app on a microbit v2 board.

### TODO or Help Wanted

Feedback is much appreciated.

### Documentation Updated

- [x] no updates are required.

### CODE REVIEW

**Review by alexandruradovici — COMMENTED** [review]
Looks ok.

**Comment by alexandruradovici** [comment]
Hi, any updates regarding the writing of unit tests?

**Comment by alexandruCarp** [comment]
Hi! Sorry for this long break, to be honest I didn't really know how this testing part works and postponed this for too long. I don't want to give up though, so if it's not too late I will try to get it done by the end of this week. I guess taking the tests for leds/buttons APIs as an example should be ok, right?

**Comment by alexandruradovici** [comment]
Yes, looking at LED and GPIO should be the staring point. Briefly, you should:
- write a fake temperature driver 
- write tests for it
- write tests that use the fake kernel and driver to test your user space driver library

**Comment by alexandruCarp** [comment]
I don't know how to test the "read_temperature_sync()" function as I would have to schedule an upcall after the "read_temp" command, but before yield, and they both happen in that function. I thought that maybe I should modify the fake driver to schedule an upcall on the call of the command, but it doesn't seem right for the upcall to be scheduled immediately (the driver would never be busy). Any suggestions?

**Comment by alexandruradovici** [comment]
The `sync` function should return the temperature, as simple as:

```rust
assert_eq! (Temperature::read_temperature_sync(), Ok(expected_temperature_value));
```

File: apis/temperature/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    /// Unregister the events listener
+    pub fn unregister_listener() {
+        S::unsubscribe(DRIVER_NUM, 0)
+    }
+
+    /// initiate a syncronous temperature mesurement
+    /// Returns Ok(temperature_value) if the operation was successful
+    pub fn read_temperature_sync() -> Result<u32, ErrorCode> {
```
Comment (alexandruradovici) [inline apis/temperature/src/lib.rs:35]:
```suggestion
    pub fn read_temperature_sync() -> Result<isize, ErrorCode> {
```

The temperature is reported in 100s of centigrade, meaning that 20.6 deg is 2060. The temperature can be negative, but as upcall values are always `usize`, the driver makes a typecast

https://github.com/tock/tock/blob/500caa5e882b7efe40158849fb489205f07c6a2e/capsules/src/temperature.rs#L119

that the library should reverse.

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

**Comment by alexandruCarp** [comment]
> The `sync` function should return the temperature, as simple as:
> 
> ```rust
> assert_eq! (Temperature::read_temperature_sync(), Ok(expected_temperature_value));
> ```

Right, and it works as expected with the real driver on the board. However, with the fake driver it says `yield-wait called with no queued upcall` because if I call the [set_value](https://github.com/alexandruCarp/libtock-rs/blob/ad9b1a1e0f840045551d61b29e287854253ccdfb/unittest/src/fake/temperature/mod.rs#L23) function before read_sync it does not enqueue any upcall as no subscribe has been made yet. Most likely the fake driver has a problem.

File: unittest/src/fake/temperature/tests.rs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        );
+
+        temp.set_value(200);
+        assert_eq!(fake::Syscalls::yield_no_wait(), YieldNoWaitReturn::NoUpcall);
+
+        assert!(fake::Syscalls::command(DRIVER_NUM, READ_TEMP, 0, 1).is_success());
+        temp.set_value(200);
+        assert_eq!(fake::Syscalls::yield_no_wait(), YieldNoWaitReturn::Upcall);
+    });
+}
```
Comment (alexandruradovici) [inline unittest/src/fake/temperature/tests.rs:61]: Add newline

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

**Comment by alexandruradovici** [comment]
> Right, and it works as expected with the real driver on the board. However, with the fake driver it says `yield-wait called with no queued upcall` because if I call the [set_value](https://github.com/alexandruCarp/libtock-rs/blob/ad9b1a1e0f840045551d61b29e287854253ccdfb/unittest/src/fake/temperature/mod.rs#L23) function before read_sync it does not enqueue any upcall as no subscribe has been made yet. Most likely the fake driver has a problem.

I think you have to add a different `set_value` function, like `set_value_sync` that does not call an upcall immediately, but calls it when it receives the read command.

File: unittest/src/fake/temperature/mod.rs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+//! a function `set_value` used to immediately call an upcall with a temperature value read by the sensor
+//! and a function 'set_value_sync' used to call the upcall when the read command is received.
+
+use crate::{DriverInfo, DriverShareRef};
+use libtock_platform::{CommandReturn, ErrorCode};
+use std::cell::Cell;
+
+pub struct Temperature {
+    busy: Cell<bool>,
+    upcall_on_command: Cell<Option<i32>>,
```
Comment (alexandruradovici) [inline unittest/src/fake/temperature/mod.rs:14]: Please add a comment explaining why this is required, I assume others will have this issue.

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

File: apis/temperature/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    ) -> Result<(), ErrorCode> {
+        S::subscribe::<_, _, DefaultConfig, DRIVER_NUM, 0>(subscribe, listener)
+    }
+
+    /// Unregister the events listener
+    pub fn unregister_listener() {
+        S::unsubscribe(DRIVER_NUM, 0)
+    }
+
+    /// initiate a syncronous temperature mesurement
```
Comment (alexandruradovici) [inline apis/temperature/src/lib.rs:35]:
```suggestion
    /// Initiate a syncronous temperature mesurement.
```

File: apis/temperature/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+pub struct Temperature<S: Syscalls>(S);
+
+impl<S: Syscalls> Temperature<S> {
+    /// Returns Ok() if the driver was present.This does not necessarily mean
+    /// that the driver is working.
+    pub fn exists() -> Result<(), ErrorCode> {
+        S::command(DRIVER_NUM, EXISTS, 0, 0).to_result()
+    }
+
+    /// initiate a temperature measurement used both for syncronous and asyncronous readings
```
Comment (alexandruradovici) [inline apis/temperature/src/lib.rs:17]:
```suggestion
    /// Initiate a temperature measurement.
    ///
    /// This function is used both for synchronous and asynchronous readings
```

File: apis/temperature/src/lib.rs (hunk 4 above)
Comment (alexandruradovici) [reply apis/temperature/src/lib.rs:35]: Please mention the measuring unit that the temperature is returned in (hundreds of centigrades).

File: examples/temperature.rs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        Err(_) => {
+            writeln!(Console::writer(), "temperature driver unavailable").unwrap();
+            return;
+        }
+    }
+
+    loop {
+        match Temperature::read_temperature_sync() {
+            Ok(temp_val) => {
+                writeln!(Console::writer(), "Temperature: {}*C\n", temp_val / 100).unwrap()
```
Comment (alexandruradovici) [inline examples/temperature.rs:29]:
```suggestion
                writeln!(Console::writer(), "Temperature: {}.{}*C\n", temp_val / 100, temp_val % 100).unwrap()
```

**Review by alexandruradovici — CHANGES_REQUESTED** [review]
Write a comment with the pull request message (use the pull request template from Tock) and this should be ready for sending to upstream Tock.

**Comment by alexandruradovici** [comment]
Please run `cargo fmt` to format the source.

**Comment by alexandruCarp** [comment]
> Write a comment with the pull request message (use the pull request template from Tock) and this should be ready for sending to upstream Tock.

### Pull Request Overview

This pull request adds:
- a Temperature API, similar to the [libtock-c version](https://github.com/tock/libtock-c/blob/master/libtock/temperature.h). The API provides a function that simulates a synchronous temperature reading and also functions to initiate a reading, register/ unregister a listener and check for driver existence. 
- an example app which samples the temperature sensor every 2 seconds and prints the measured temperature.
- a `fake` temperature driver, for testing purposes.
- unit tests for the fake driver and for the user space library

### Testing Strategy

This pull request was tested with the unit tests and by running the example app on a microbit v2 board.

### TODO or Help Wanted

Feedback is much appreciated.

### Documentation Updated

- [x] no updates are required.

### Formatting

- [ ] Ran `make prepush`.

**Comment by alexandruCarp** [comment]
Should I open a pull request to upstream Tock now?

**Comment by alexandruradovici** [comment]
Please fix the clippy issue and after that you can send the pull request.

**Comment by alexandruCarp** [comment]
Opened PR tock/libtock-rs#451.

**Comment by alexandruradovici** [comment]
Please run `make prepush` and if it works, check the `ran make prepush` checkbox.

**Comment by alexandruCarp** [comment]
> Please run `make prepush` and if it works, check the `ran make prepush` checkbox in both PRs.

There is no `prepush` rule in [libtock-rs/Makefile](https://github.com/UPB-CS-OpenSourceUpstream/libtock-rs/blob/master/Makefile). Am I missing something?

**Comment by alexandruradovici** [comment]
> > Please run `make prepush` and if it works, check the `ran make prepush` checkbox in both PRs.
> 
> There is no `prepush` rule in [libtock-rs/Makefile](https://github.com/UPB-CS-OpenSourceUpstream/libtock-rs/blob/master/Makefile). Am I missing something?

You are right, there is no PR template. Please just delete that form the pull request message.

**Review by alexandruradovici — DISMISSED** [review]

**Comment by alexandruradovici** [comment]
@alexandruCarp Please select a timeslot for the interview https://doodle.com/meeting/participate/id/eZ8gOOEb.

**Comment by alexandruradovici** [comment]
Merged in upstream.

**Pull request closed without merging**